src/platform/P6/OTAImageProcessorImpl.cpp

```diff
@@ -0,0 +1,222 @@
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    else if (imageProcessor->mDownloader == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+
+    // Open and erase secondary flash area to prepare
+    if (flash_area_open(FLASH_AREA_IMAGE_SECONDARY(0), &(imageProcessor->mFlashArea)) != 0)
+    {
+        imageProcessor->mDownloader->OnPreparedForDownload(CHIP_ERROR_OPEN_FAILED);
+        return;
+    }
+    if (flash_area_erase(imageProcessor->mFlashArea, 0, imageProcessor->mFlashArea->fa_size) != 0)
+    {
+        imageProcessor->mDownloader->OnPreparedForDownload(CHIP_ERROR_OPEN_FAILED);
+        return;
+    }
+
+    imageProcessor->mDownloader->OnPreparedForDownload(CHIP_NO_ERROR);
+}
+
+void OTAImageProcessorImpl::HandleFinalize(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    flash_area_close(imageProcessor->mFlashArea);
+    ChipLogProgress(SoftwareUpdate, "Setting boot pending");
+    int ret = boot_set_pending(0, 1);
+
+    if (ret != 0)
+    {
+        ChipLogError(SoftwareUpdate, "Failed to set boot pending");
+        return;
+    }
+
+    imageProcessor->ReleaseBlock();
+}
+
+void OTAImageProcessorImpl::HandleAbort(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    flash_area_erase(imageProcessor->mFlashArea, 0, imageProcessor->mFlashArea->fa_size);
+
+    imageProcessor->ReleaseBlock();
+}
+
+void OTAImageProcessorImpl::HandleProcessBlock(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    else if (imageProcessor->mDownloader == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+
+    int rc = flash_area_write(imageProcessor->mFlashArea, imageProcessor->mParams.downloadedBytes, imageProcessor->mBlock.data(),
+                              imageProcessor->mBlock.size());
+    if (rc != 0)
+    {
+        imageProcessor->mDownloader->EndDownload(CHIP_ERROR_WRITE_FAILED);
+        return;
+    }
+
+    imageProcessor->mParams.downloadedBytes += imageProcessor->mBlock.size();
+    imageProcessor->mDownloader->FetchNextData();
+}
+
+void OTAImageProcessorImpl::HandleApply(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    chip::OTARequestorInterface * requestor = chip::GetRequestorInstance();
+    if (requestor != nullptr)
+    {
+        requestor->NotifyUpdateApplied(CHIP_DEVICE_CONFIG_DEVICE_SOFTWARE_VERSION + 1);
```

Comment:
This should be done after the reboot, no? At this stage, the update is about to be applied?

Also, does `CHIP_DEVICE_CONFIG_DEVICE_SOFTWARE_VERSION` get updated dynamically?

Reply:
So I've moved this to check OTAUpdateState after boot and call notify if it's a fresh update. I'm not sure if that OTA state is being saved out persistently though. Some of the discussions on Slack as well as this issue https://github.com/project-chip/connectedhomeip/issues/14338 makes me think that persistent OTA storage is not implemented yet. Is that correct?

Reply:
Correct, persistent storage is being worked on as we speak. However, note that software version is not something that will be persisted in OTA class. This should be done per platform. That should be the value used here.

Reply:
Ok I've reworked things so that this checks the persisted version against CHIP_DEVICE_CONFIG_DEVICE_SOFTWARE_VERSION (whose value will be based on whether it's an upgrade image or not). One thing to note since: I did have to add a declaration for StoreSoftwareVersion to the ConfigurationManager interface (src/include/platform/ConfigurationManager.h).